### PR TITLE
Register a CRD for v1beta1 PeerAuthentication

### PR DESCRIPTION
**Pull request by diemtvu**
- Update to latest API @7ecef25
- Add new CRD for `PeerAuthentication` and related validation.
- Populate `PeerAuthentication` to push context.

Issue: https://github.com/istio/istio/issues/20746

### CODE REVIEW

File: pilot/pkg/model/authentication.go [hunk 1]
```diff
@@ -77,7 +80,11 @@ func initAuthenticationPolicies(env *Environment) *AuthenticationPolicies {
 		policy.addRequestAuthentication(configs)
 	}
 
-	// TODO(diemtvu): populate mTLS mode from mesh config and namespace labels.
+	if configs, err := env.List(collections.IstioSecurityV1Beta1Peerauthentications.Resource().GroupVersionKind(), NamespaceAll); err == nil {
```
Comment (incfly) [inline pilot/pkg/model/authentication.go]: i'll suggest to break the line length. also we should log if error occurs

File: pilot/pkg/model/authentication.go [hunk 2]
```diff
@@ -88,27 +95,60 @@ func (policy *AuthenticationPolicies) addRequestAuthentication(configs []Config)
...
+	}
+	return getConfigsForWorkload(policy.peerAuthentications, policy.rootNamespace, extractor, namespace, workloadLabels)
+}
+
+// GetRootNamespace return root namespace that is tracked by the policy object.
+func (policy *AuthenticationPolicies) GetRootNamespace() string {
+	return policy.rootNamespace
+}
+
+type extractSelector func(*Config) map[string]string
```
Comment (incfly) [inline pilot/pkg/model/authentication.go]: func comment for type defintion, what map[string]string stands for

File: pilot/pkg/model/authentication.go [hunk 3]
```diff
@@ -88,27 +95,60 @@ func (policy *AuthenticationPolicies) addRequestAuthentication(configs []Config)
...
+		policy.peerAuthentications[config.Namespace] =
+			append(policy.peerAuthentications[config.Namespace], config)
+	}
+}
+
 // GetJwtPoliciesForWorkload returns a list of JWT policies matching to labels.
 func (policy *AuthenticationPolicies) GetJwtPoliciesForWorkload(namespace string,
+	workloadLabels labels.Collection) []*Config {
+	extractor := func(cfg *Config) map[string]string {
+		return cfg.Spec.(*v1beta1.RequestAuthentication).GetSelector().GetMatchLabels()
```
Comment (incfly) [inline pilot/pkg/model/authentication.go]: We should be able to just do type switch of the opaque config, and then get label from config spec.

this collapse the oneliner funtion into a consolidated place, easier to follow.

File: pilot/pkg/model/authentication.go [hunk 4]
```diff
@@ -68,6 +70,7 @@ type AuthenticationPolicies struct {
 func initAuthenticationPolicies(env *Environment) *AuthenticationPolicies {
 	policy := &AuthenticationPolicies{
 		requestAuthentications: map[string][]Config{},
+		peerAuthentications:    map[string][]Config{},
 		namespaceMTLSMode:      map[string]MutualTLSMode{},
```
Comment (incfly) [inline pilot/pkg/model/authentication.go]: namespacemTLSMode, obsolete?

**Review by incfly — COMMENTED** [review]
will be better to make api dep update and actual code change in separate PR. so what is changed becomes more obvious.

File: pilot/pkg/model/authentication.go (hunk 3 above)
Comment (diemtvu) [reply pilot/pkg/model/authentication.go]: Done

File: pilot/pkg/model/authentication.go (hunk 2 above)
Comment (diemtvu) [reply pilot/pkg/model/authentication.go]: Removed.

File: pilot/pkg/model/authentication.go (hunk 1 above)
Comment (diemtvu) [reply pilot/pkg/model/authentication.go]: Done. Also return error back to caller.

**Review by diemtvu — COMMENTED** [review]
> will be better to make api dep update and actual code change in separate PR. so what is changed becomes more obvious.

I intended to do so, but then need to verify the API gen files was correct (last time I miss some annotations and didn't cause problem till we start adding the CRD). I can move the push context out of this PR, though at this point, it may not add much value.

**Comment by diemtvu** [comment]
> > will be better to make api dep update and actual code change in separate PR. so what is changed becomes more obvious.
> 
> I intended to do so, but then need to verify the API gen files was correct (last time I miss some annotations and didn't cause problem till we start adding the CRD). I can move the push context out of this PR, though at this point, it may not add much value.

It's actually cleaner now after rebase, as someone already update API for me :).

**Review by yangminzhu — APPROVED** [review]

**Review by howardjohn — APPROVED** [review]

File: pilot/pkg/model/authentication.go [hunk 5]
```diff
@@ -57,28 +57,37 @@ type AuthenticationPolicies struct {
...
+		peerAuthentications:    map[string][]Config{},
 		rootNamespace:          env.Mesh().GetRootNamespace(),
 	}
 
-	if configs, err := env.List(collections.IstioSecurityV1Beta1Requestauthentications.Resource().GroupVersionKind(), NamespaceAll); err == nil {
+	if configs, err := env.List(
+		collections.IstioSecurityV1Beta1Requestauthentications.Resource().GroupVersionKind(), NamespaceAll); err == nil {
 		sortConfigByCreationTime(configs)
 		policy.addRequestAuthentication(configs)
+	} else {
```
Comment (incfly) [inline pilot/pkg/model/authentication.go]: if err != nil, return earlier can reduce the nesting level.

File: pilot/pkg/model/authentication.go [hunk 6]
```diff
@@ -57,28 +57,37 @@ type AuthenticationPolicies struct {
...
+		return nil, err
 	}
 
-	// TODO(diemtvu): populate mTLS mode from mesh config and namespace labels.
-	return policy
+	if configs, err := env.List(
+		collections.IstioSecurityV1Beta1Peerauthentications.Resource().GroupVersionKind(), NamespaceAll); err == nil {
+		sortConfigByCreationTime(configs)
+		policy.addPeerAuthentication(configs)
+	} else {
```
Comment (incfly) [inline pilot/pkg/model/authentication.go]: same here.

**Review by incfly — COMMENTED** [review]
lgtm, a few nitpicks.

**Pull request merged**